File: smalivm/src/main/java/org/cf/smalivm/opcode/UnaryMathOp.java
package org.cf.smalivm.opcode;

import org.cf.smalivm.context.ExecutionNode;
import org.cf.smalivm.context.HeapItem;
import org.cf.smalivm.context.MethodState;
import org.cf.util.Utils;
import org.jf.dexlib2.builder.MethodLocation;

public class UnaryMathOp extends MethodStateOp {

    private final int destRegister;
    private final int srcRegister;

    UnaryMathOp(MethodLocation location, MethodLocation child, int destRegister, int srcRegister) {
        super(location, child);

        this.destRegister = destRegister;
        this.srcRegister = srcRegister;
    }

    private static String getResultTypeName(String opName) {
        String[] parts = opName.split("-");
        String type = parts[parts.length - 1];
        String resultType = null;
        if ("int".equals(type)) {
            resultType = "I";
        } else if ("long".equals(type)) {
            resultType = "J";
        } else if ("float".equals(type)) {
            resultType = "F";
        } else if ("byte".equals(type)) {
            resultType = "B";
        } else if ("char".equals(type)) {
            resultType = "C";
        } else if ("double".equals(type)) {
            resultType = "D";
        } else if ("short".equals(type)) {
            resultType = "S";
        }

        return resultType;
    }

    public int getDestRegister() {
        return destRegister;
    }

    public int getSrcRegister() {
        return srcRegister;
    }

    @Override
    public void execute(ExecutionNode node, MethodState mState) {
        HeapItem item = mState.readRegister(srcRegister);
        HeapItem resultItem;
        String type = getResultTypeName(getName());
        if (item.isUnknown()) {
            resultItem = HeapItem.newUnknown(type);
        } else {
            Object resultValue = perform(item.getValue(), getName());
            resultItem = new HeapItem(resultValue, type);
        }
        mState.assignRegister(destRegister, resultItem);
    }

    @Override
    public String toString() {
        return getName() + " r" + destRegister + ", r" + srcRegister;
    }

    private Object perform(Object value, String opName) {
        // Stupid, but simple. Deals.
        Object resultValue = null;
        if (opName.startsWith("double")) {
            Double typedValue = Utils.getDoubleValue(value);
            if (opName.endsWith("float")) {
                resultValue = typedValue.floatValue();
            } else if (opName.endsWith("int")) {
                resultValue = typedValue.intValue();
            } else if (opName.endsWith("long")) {
                resultValue = typedValue.longValue();
            }
        } else if (opName.startsWith("float")) {
            Float typedValue = Utils.getFloatValue(value);
            if (opName.endsWith("double")) {
                resultValue = typedValue.doubleValue();
            } else if (opName.endsWith("int")) {
                resultValue = typedValue.intValue();
            } else if (opName.endsWith("long")) {
                resultValue = typedValue.longValue();
            }
        } else if (opName.startsWith("long")) {
            Long typedValue = Utils.getLongValue(value);
            if (opName.endsWith("double")) {
                resultValue = typedValue.doubleValue();
            } else if (opName.endsWith("int")) {
                resultValue = typedValue.intValue();
            } else if (opName.endsWith("float")) {
                resultValue = typedValue.floatValue();
            }
        } else if (opName.startsWith("int")) {
            // Could be something other than an int, such as short
            Integer typedValue = Utils.getIntegerValue(value);
            if (opName.endsWith("byte")) {
                resultValue = typedValue.byteValue();
            } else if (opName.endsWith("char")) {
                resultValue = (char) typedValue.intValue();
            } else if (opName.endsWith("short")) {
                resultValue = typedValue.shortValue();
            } else if (opName.endsWith("double")) {
                resultValue = typedValue.doubleValue();
            } else if (opName.endsWith("long")) {
                resultValue = typedValue.longValue();
            } else if (opName.endsWith("float")) {
                resultValue = typedValue.floatValue();
            }
        } else if (opName.startsWith("neg")) {
            if (opName.endsWith("double")) {
                Double typedValue = Utils.getDoubleValue(value);
                resultValue = -typedValue;
            } else if (opName.endsWith("float")) {
                Float typedValue = Utils.getFloatValue(value);
                resultValue = -typedValue;
            } else if (opName.endsWith("int")) {
                Integer typedValue = Utils.getIntegerValue(value);
                resultValue = -typedValue;
            } else if (opName.endsWith("long")) {
                Long typedValue = Utils.getLongValue(value);
                resultValue = -typedValue;
            }
        } else if (opName.startsWith("not")) {
            if (opName.endsWith("int")) {
                Integer typedValue = Utils.getIntegerValue(value);
                resultValue = ~typedValue;
            } else if (opName.endsWith("long")) {
                Long typedValue = Utils.getLongValue(value);
                resultValue = ~typedValue;
            }
        }

        return resultValue;
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/UnaryMathOp.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.configuration.Configuration
import org.cf.smalivm.dex.CommonTypes
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm2.ExecutionNode
import org.cf.smalivm2.UnresolvedChild
import org.cf.smalivm2.Value
import org.cf.util.Utils
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.formats.Instruction12x

class UnaryMathOp internal constructor(location: MethodLocation, child: MethodLocation, val destRegister: Int, val srcRegister: Int) :
    Op(location, child) {

    override val registersReadCount = 1
    override val registersAssignedCount = 1

    override fun execute(node: ExecutionNode): Array<out UnresolvedChild> {
        val item = node.state.readRegister(srcRegister)
        val result: Value
        val type = getResultTypeName(name)
        result = if (item.isUnknown) {
            Value.unknown(type)
        } else {
            val resultValue = perform(item, name)
            Value.wrap(resultValue, type)
        }
        node.state.assignRegister(destRegister, result)
        return finishOp()
    }

    override fun toString() = "$name r$destRegister, r$srcRegister"

    private fun perform(value: Value, opName: String): Any {
        return when {
            opName.startsWith("double") -> {
                val typedValue = value.toDouble()
                when {
                    opName.endsWith("float") -> typedValue.toFloat()
                    opName.endsWith("long") -> typedValue.toLong()
                    opName.endsWith("int") -> typedValue.toInt()
                    else -> throw IllegalArgumentException("Unexpected op name $opName")
                }
            }
            opName.startsWith("float") -> {
                val typedValue = value.toFloat()
                when {
                    opName.endsWith("double") -> typedValue.toDouble()
                    opName.endsWith("long") -> typedValue.toLong()
                    opName.endsWith("int") -> typedValue.toInt()
                    else -> throw IllegalArgumentException("Unexpected op name $opName")
                }
            }
            opName.startsWith("long") -> {
                val typedValue = value.toLong()
                when {
                    opName.endsWith("double") -> typedValue.toDouble()
                    opName.endsWith("float") -> typedValue.toFloat()
                    opName.endsWith("int") -> typedValue.toInt()
                    else -> throw IllegalArgumentException("Unexpected op name $opName")
                }
            }
            opName.startsWith("int") -> {
                // Could be something other than an int, such as short
                val typedValue = value.toInteger()
                when {
                    opName.endsWith("byte") -> typedValue.toByte()
                    opName.endsWith("char") -> typedValue.toChar()
                    opName.endsWith("short") -> typedValue.toShort()
                    opName.endsWith("double") -> typedValue.toDouble()
                    opName.endsWith("long") -> typedValue.toLong()
                    opName.endsWith("float") -> typedValue.toFloat()
                    else -> throw IllegalArgumentException("Unexpected op name $opName")
                }
            }
            opName.startsWith("neg") -> {
                when {
                    opName.endsWith("double") -> -value.toDouble()
                    opName.endsWith("float") -> -value.toFloat()
                    opName.endsWith("long") -> -value.toLong()
                    opName.endsWith("int") -> -value.toInteger()
                    else -> throw IllegalArgumentException("Unexpected op name $opName")
                }
            }
            opName.startsWith("not") -> {
                when {
                    opName.endsWith("int") -> value.toInteger().inv()
                    opName.endsWith("long") -> value.toLong().inv()
                    else -> throw IllegalArgumentException("Unexpected op name $opName")
                }
            }
            else -> throw IllegalArgumentException("Unexpected op name $opName")
        }
    }

    companion object : OpFactory {
        private fun getResultTypeName(opName: String): String {
            val parts = opName.split("-").toTypedArray()
            return when (parts[parts.size - 1]) {
                "int" -> CommonTypes.INTEGER
                "long" -> CommonTypes.LONG
                "float" -> CommonTypes.FLOAT
                "byte" -> CommonTypes.BYTE
                "char" -> CommonTypes.CHARACTER
                "double" -> CommonTypes.DOUBLE
                "short" -> CommonTypes.SHORT
                else -> throw IllegalArgumentException("Unexpected op name $opName")
            }
        }

        override fun build(
            location: MethodLocation,
            addressToLocation: Map<Int, MethodLocation>,
            classManager: ClassManager,
            classLoader: SmaliClassLoader,
            configuration: Configuration
        ): Op {
            val child = Utils.getNextLocation(location, addressToLocation)
            val instr = location.instruction as Instruction12x
            val destRegister = instr.registerA
            val srcRegister = instr.registerB
            return UnaryMathOp(location, child, destRegister, srcRegister)
        }
    }
}
